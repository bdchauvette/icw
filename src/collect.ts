export async function* collect<T>(
  iterable: AsyncIterable<T> | Iterable<T>
): AsyncIterable<T[]> {
  let collectedResults = [];
  for await (let result of iterable) collectedResults.push(result);
  yield collectedResults;
}
