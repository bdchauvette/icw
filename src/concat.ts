import { flat } from "./flat";
import { toIterable } from "./__internal__/toIterable";
import { IterableLike } from "./IterableLike";

export async function* concat<T, U>(
  iterableLike: IterableLike<T>,
  ...values: U[]
): AsyncIterableIterator<T | U> {
  yield* toIterable(iterableLike);
  yield* flat(values, 1);
}
