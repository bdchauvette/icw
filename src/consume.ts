export const consume = async <T>(
  iterable: Iterable<T> | AsyncIterable<T>
): Promise<void> => {
  // @ts-ignore
  for await (let result of iterable) continue;
};
