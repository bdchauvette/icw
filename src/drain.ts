export async function drain<T>(
  iterable: AsyncIterable<T> | Iterable<T>
): Promise<void> {
  // @ts-ignore
  for await (let result of iterable) continue;
}
