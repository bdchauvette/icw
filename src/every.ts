import { IterableLike } from "./IterableLike";
import { withIndex } from "./withIndex";

export async function every<T>(
  iterableLike: IterableLike<T>,
  predicate: (value: T, index?: number) => boolean | Promise<boolean>,
  thisArg?: any
): Promise<boolean> {
  let iterableDidPass = false;

  for await (let [value, index] of withIndex(iterableLike)) {
    iterableDidPass = Reflect.apply(predicate, thisArg, [value, index]);
    if (!(await iterableDidPass)) return false;
  }

  return iterableDidPass;
}
