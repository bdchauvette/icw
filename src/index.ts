export { ICW } from "./ICW";

export { from } from "./from";
export { of } from "./of";

export { drain } from "./drain";
export { filter } from "./filter";
export { map } from "./map";
export { reject } from "./reject";
export { scan } from "./scan";
export { withIndex } from "./withIndex";
