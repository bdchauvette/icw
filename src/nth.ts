import { isNumber } from "./__internal__/isNumber";
import { isString } from "./__internal__/isString";
import { withIndex } from "./withIndex";
import { IterableLike } from "./IterableLike";

export async function nth<T>(
  iterableLike: IterableLike<T>,
  targetIndex: number
): Promise<T | undefined> {
  if (!isNumber(targetIndex) || targetIndex < 0) {
    throw new RangeError("targetIndex must be a non-negative number");
  }

  // Avoid iterating if we can do random access
  if (Array.isArray(iterableLike) || isString(iterableLike)) {
    return iterableLike[targetIndex];
  }

  for await (let [value, resultIndex] of withIndex(iterableLike)) {
    if (resultIndex === targetIndex) return value;
  }
}
