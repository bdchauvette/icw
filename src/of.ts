/**
 * Creates an AsyncIterator that yields each argument.
 */
export function of<T>(...values: T[]): AsyncIterable<T> {
  return {
    async *[Symbol.asyncIterator](): AsyncIterableIterator<T> {
      yield* values;
    }
  };
}
