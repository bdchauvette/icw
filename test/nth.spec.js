import { nth } from "../src/nth";

import { of } from "../src/of";
import { ArrayLike } from "./helpers/ArrayLike";

test("rejects on non-IterableLike input", async () => {
  expect.assertions(2);
  try {
    await nth(null, 0);
  } catch (error) {
    expect(error).toBeInstanceOf(TypeError);
    expect(error.message).toMatchInlineSnapshot(
      `"Must provide an iterable, async iterable, Array-like value, or a Promise."`
    );
  }
});

test.each`
  description      | targetIndex
  ${"null"}        | ${null}
  ${"undefined"}   | ${undefined}
  ${"non-numeric"} | ${"-1"}
  ${"< 0"}         | ${-1}
`("rejects when targetIndex is $description", async ({ targetIndex }) => {
  expect.assertions(2);
  try {
    await nth(of("foo"), targetIndex);
  } catch (error) {
    expect(error).toBeInstanceOf(RangeError);
    expect(error.message).toMatchSnapshot();
  }
});

test("eagerly consumes wrapped IterableLike input", async () => {
  expect.assertions(1);
  await expect(_ => nth(_, 1)).toEagerlyConsumeWrappedAsyncIterable();
});

test.each`
  inputType          | iterableLike                          | index | expectedValue
  ${"AsyncIterable"} | ${of("foo", "bar", "baz")}            | ${1}  | ${"bar"}
  ${"Iterable"}      | ${["foo", "bar", "baz"]}              | ${1}  | ${"bar"}
  ${"ArrayLike"}     | ${new ArrayLike("foo", "bar", "baz")} | ${1}  | ${"bar"}
  ${"Promise"}       | ${Promise.resolve("foo")}             | ${0}  | ${"foo"}
`(
  "resolves to the nth value from $inputType input",
  async ({ iterableLike, index, expectedValue }) => {
    expect.assertions(1);
    await expect(nth(iterableLike, index)).resolves.toStrictEqual(
      expectedValue
    );
  }
);

test.each`
  inputType          | iterableLike
  ${"AsyncIterable"} | ${of("foo", "bar", "baz")}
  ${"Iterable"}      | ${["foo", "bar", "baz"]}
  ${"ArrayLike"}     | ${new ArrayLike("foo", "bar", "baz")}
  ${"Promise"}       | ${Promise.resolve("foo")}
`(
  "resolves to undefined if $inputType input does not have an nth value",
  async ({ iterableLike }) => {
    expect.assertions(1);
    await expect(
      nth(iterableLike, Number.MAX_SAFE_INTEGER)
    ).resolves.toBeUndefined();
  }
);
