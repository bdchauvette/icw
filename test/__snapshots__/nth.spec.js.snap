// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rejects when targetIndex is "< 0" 1`] = `"targetIndex must be a non-negative number"`;

exports[`rejects when targetIndex is "non-numeric" 1`] = `"targetIndex must be a non-negative number"`;

exports[`rejects when targetIndex is "null" 1`] = `"targetIndex must be a non-negative number"`;

exports[`rejects when targetIndex is "undefined" 1`] = `"targetIndex must be a non-negative number"`;
